### PR TITLE
Make tutorial self-sufficient

### PR DESCRIPTION
**Pull request by TobiasBabin**
This PR makes the repo self-sufficient in the way that users of the "[Deploy ephemeral environments](https://developer.humanitec.com/introduction/tutorials/deploy-ephemeral-environments/)" tutorial, which will be updated in parallel, can perform a complete life cycle including

- building and deploying a container image
- branching, creating a PR and having an ephemeral environment created
- applying modifications to the code
- previewing modifications in the ephemeral env
- finally merging modifications

The PR leaves the GitHub workflows largely unchanged except adding Docker build+push stages.

The application itself is a minimalistic Node.js express app.

### CODE REVIEW

File: .github/workflows/pull_request.yaml [hunk 1]
```diff
@@ -2,57 +2,84 @@ name: Pull Request
...
+          IMAGE=ghcr.io/$TAG
+
+          # Set this label to connect the package to the GitHub repo to make it public
+          docker build . --tag $IMAGE  \
+            --label "org.opencontainers.image.source=https://github.com/$GH_REPOSITORY"
+          docker push $IMAGE
+          echo "IMAGE=$IMAGE" >> $GITHUB_OUTPUT
       - name: Run Score
         run: |
           humctl score deploy \
```
Comment (mathieu-benoit) [inline .github/workflows/pull_request.yaml:68]: Let's add an explicit message/comment like this below?
```
--message "$(curl -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" "${{ github.event.pull_request.commits_url }}?per_page=10" | jq -r .[-1].commit.message)" \
```

File: .github/workflows/main.yaml [hunk 2]
```diff
@@ -3,26 +3,48 @@ on:
...
+
+          # Set this label to connect the package to the GitHub repo to make it public
+          docker build . --tag $IMAGE  \
+            --label "org.opencontainers.image.source=https://github.com/$GH_REPOSITORY"
+          docker push $IMAGE
+          echo "IMAGE=$IMAGE" >> $GITHUB_OUTPUT
+
+      - name: Deploy Score
         run: |
           humctl score deploy \
```
Comment (mathieu-benoit) [inline .github/workflows/main.yaml:46]: Let's add an explicit message/comment here too?

File: .github/workflows/pull_request.yaml [hunk 3]
```diff
@@ -2,57 +2,84 @@ name: Pull Request
...
+            -t ${{ env.ENVIRONMENT_TYPE }} \
+            --from ${{ env.BASE_ENVIRONMENT }}
+        
+      - name: Build and push image
+        id: build_push
+        run: |
+          # Make sure repo name is all lowercase for docker push
+          GH_REPOSITORY=$(echo ${{ github.repository }} | tr '[:upper:]' '[:lower:]')
+
+          TAG=$GH_REPOSITORY:${{ env.ENVIRONMENT_ID }}
```
Comment (mathieu-benoit) [inline .github/workflows/pull_request.yaml:55]: Depending on the `imagePullPolicy` in the cluster, if multiple images are built via this PR, the updates in this image may not be reflected in the running container... not a big deal as we are mostly illustrating Score/Humanitec... but just for your info.
Another way could be to still keep this tag, and in Score passing the `digest`, something around this:
```
oras manifest fetch ${{ env.IMAGE_NAME }}:${{ env.IMAGE_TAG }} \
              --descriptor \
              | jq -r .digest
```

**Review by mathieu-benoit — APPROVED** [review]
3 comments, but even without them taken into account, LGTM

File: .github/workflows/main.yaml (hunk 2 above)
Comment (TobiasBabin) [reply .github/workflows/main.yaml:46]: Very good point. I added a message showing the latest git commit message. This works (I tested it) without the curl to the GH API, preventing a dependency to GitHub for any adopters.

File: .github/workflows/pull_request.yaml (hunk 1 above)
Comment (TobiasBabin) [reply .github/workflows/pull_request.yaml:68]: I added a message showing the latest git commit message. This works (I tested it) without the curl to the GH API, preventing a dependency to GitHub for any adopters. I needed to change the `actions/checkout@v4` to use

```yaml
      - uses: actions/checkout@v4
        with:
          # Checkout using the head commit to read commit message further down
          ref: ${{ github.event.pull_request.head.sha }}
```

based on [this discussion](https://github.com/orgs/community/discussions/28474#discussioncomment-3334578).

File: .github/workflows/pull_request.yaml (hunk 3 above)
Comment (TobiasBabin) [reply .github/workflows/pull_request.yaml:55]: Yes, good point. I'll check if I can obtain the image SHA256 and then replace `<image-name>:<tag>` with `<image-name>@<digest>` in the `score deploy`.

Comment (TobiasBabin) [reply .github/workflows/pull_request.yaml:55]: See the change below (line 62): we now get the digest via a `docker inspect` (preventing the `oras` dependency) and pass it on to the Score deploy step.

**Review by mathieu-benoit — APPROVED** [review]
LGTM

File: .github/workflows/close_pr.yaml [hunk 4]
```diff
@@ -1,17 +1,17 @@
...
-  HUMCTL_VERSION: '0.23.2'
+
 jobs:
   cleanup:
     runs-on: ubuntu-latest
     steps:
       - uses: humanitec/setup-cli-action@v1
         with:
-          version: ${{ env.HUMCTL_VERSION }}
+          version: ${{ vars.HUMCTL_VERSION }}
```
Comment (johanneswuerbach) [inline .github/workflows/close_pr.yaml]: If this is managed as variable, how can I safely test a humctl version bump? Right now I can just open a PR with the strings replaced, with a var its more complicated.

File: .github/workflows/pull_request.yaml [hunk 5]
```diff
@@ -2,57 +2,91 @@ name: Pull Request
...
         run: |
           humctl score deploy \
               -f score.yaml \
-              --extensions humanitec.score.yaml \
               --token ${{ secrets.HUMANITEC_TOKEN }} \
               --org ${{ vars.HUMANITEC_ORG }} \
               --app ${{ vars.APP_NAME }} \
               --env ${{ env.ENVIRONMENT_ID }} \
+              --message "$(git show -s --format=%s)" \
+              --property "containers.demo.image=${{ steps.build_push.outputs.IMAGE_DIGEST }}" \
```
Comment (johanneswuerbach) [inline .github/workflows/pull_request.yaml]: Nit: This could be `--image=${{ steps.build_push.outputs.IMAGE_DIGEST }}`

**Review by johanneswuerbach — COMMENTED** [review]
Two small comments

File: .github/workflows/close_pr.yaml (hunk 4 above)
Comment (TobiasBabin) [reply .github/workflows/close_pr.yaml]: Good practitioner's input, thanks - I pinned the version in the three workflows like previously.

File: .github/workflows/pull_request.yaml (hunk 5 above)
Comment (TobiasBabin) [reply .github/workflows/pull_request.yaml]: Thanks, adjusted!

**Review by johanneswuerbach — APPROVED** [review]

**Pull request merged**